File: Snowplow iOSTests/TestLogger.m
//
//  TestLogger.m
//  Snowplow-iOSTests
//
//  Copyright (c) 2013-2022 Snowplow Analytics Ltd. All rights reserved.
//
//  This program is licensed to you under the Apache License Version 2.0,
//  and you may not use this file except in compliance with the Apache License
//  Version 2.0. You may obtain a copy of the Apache License Version 2.0 at
//  http://www.apache.org/licenses/LICENSE-2.0.
//
//  Unless required by applicable law or agreed to in writing,
//  software distributed under the Apache License Version 2.0 is distributed on
//  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
//  express or implied. See the Apache License Version 2.0 for the specific
//  language governing permissions and limitations there under.
//
//  Authors: Alex Benini
//  License: Apache License Version 2.0
//

#import <XCTest/XCTest.h>
#import "SPLogger.h"

@interface MockDiagnosticLogger : NSObject
@property (nonatomic) void (^callback)(NSString *tag, NSString *message, NSError *error, NSException *exception);
@end

@implementation MockDiagnosticLogger

- (instancetype)init {
    if (self = [super init]) {
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(logDiagnosticError:) name:@"SPTrackerDiagnostic" object:nil];
    }
    return self;
}

- (void)logDiagnosticError:(NSNotification *)notification {
    NSDictionary *userInfo = [notification userInfo];
    NSString *tag = [userInfo objectForKey:@"tag"];
    NSString *message = [userInfo objectForKey:@"message"];
    NSError *error = [userInfo objectForKey:@"error"];
    NSException *exception = [userInfo objectForKey:@"exception"];
    self.callback(tag, message, error, exception);
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

@end

@interface TestLogger : XCTestCase

@end

@implementation TestLogger

- (void)setUp {
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
}

- (void)testDiagnosticTracking {
    XCTestExpectation *expectation = [XCTestExpectation new];
    MockDiagnosticLogger *diagnostic = [[MockDiagnosticLogger alloc] init];
    diagnostic.callback = ^(NSString *tag, NSString *message, NSError *error, NSException *exception) {
        XCTAssertEqualObjects(tag, NSStringFromClass(self.class));
        NSString *expectedMessage = [NSString stringWithFormat:@"Error test %d %@", 1, @12.3];
        XCTAssertEqualObjects(message, expectedMessage);
        [expectation fulfill];
    };

    SPLogTrack(nil, @"Error test %d %@", 1, @12.3);
    [self waitForExpectations:@[expectation] timeout:10];
}

- (void)testDiagnosticTrackingWithError {
    XCTestExpectation *expectation = [XCTestExpectation new];
    NSError *raisedError = [NSError errorWithDomain:NSURLErrorDomain code:400 userInfo:nil];

    MockDiagnosticLogger *diagnostic = [[MockDiagnosticLogger alloc] init];
    diagnostic.callback = ^(NSString *tag, NSString *message, NSError *error, NSException *exception) {
        XCTAssertEqualObjects(tag, NSStringFromClass(self.class));
        XCTAssertEqualObjects(message, @"Error test");
        XCTAssertEqual(error, raisedError);
        [expectation fulfill];
    };

    SPLogTrack(raisedError, @"Error test");
    [self waitForExpectations:@[expectation] timeout:10];
}

- (void)testDiagnosticTrackingWithException {
    XCTestExpectation *expectation = [XCTestExpectation new];
    NSException *raisedException = [NSException exceptionWithName:NSInvalidArgumentException reason:nil userInfo:nil];

    MockDiagnosticLogger *diagnostic = [[MockDiagnosticLogger alloc] init];
    diagnostic.callback = ^(NSString *tag, NSString *message, NSError *error, NSException *exception) {
        XCTAssertEqualObjects(tag, NSStringFromClass(self.class));
        XCTAssertEqualObjects(message, @"Exception test");
        XCTAssertEqual(exception, raisedException);
        [expectation fulfill];
    };

    SPLogTrack(raisedException, @"Exception test");
    [self waitForExpectations:@[expectation] timeout:10];
}

@end

File: Tests/TestLogger.swift
//  Copyright (c) 2013-2023 Snowplow Analytics Ltd. All rights reserved.
//
//  This program is licensed to you under the Apache License Version 2.0,
//  and you may not use this file except in compliance with the Apache License
//  Version 2.0. You may obtain a copy of the Apache License Version 2.0 at
//  http://www.apache.org/licenses/LICENSE-2.0.
//
//  Unless required by applicable law or agreed to in writing,
//  software distributed under the Apache License Version 2.0 is distributed on
//  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
//  express or implied. See the Apache License Version 2.0 for the specific
//  language governing permissions and limitations there under.

import XCTest
@testable import SnowplowTracker

class MockDiagnosticLogger: NSObject {
    var callback: ((_ tag: String?, _ message: String?, _ error: Error?, _ exception: NSException?) -> Void)?

    override init() {
        super.init()
        NotificationCenter.default.addObserver(self, selector: #selector(logDiagnosticError(_:)), name: NSNotification.Name("SPTrackerDiagnostic"), object: nil)
    }

    @objc func logDiagnosticError(_ notification: Notification?) {
        let userInfo = notification?.userInfo
        let tag = userInfo?["tag"] as? String
        let message = userInfo?["message"] as? String
        let error = userInfo?["error"] as? Error
        let exception = userInfo?["exception"] as? NSException
        callback?(tag, message, error, exception)
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}

class TestLogger: XCTestCase {
    override func setUp() {
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
    }

    func testDiagnosticTracking() {
        let expectation = XCTestExpectation()
        let diagnostic = MockDiagnosticLogger()
        diagnostic.callback = { tag, message, error, exception in
            XCTAssertEqual(tag, String(describing: TestLogger.self))
            let expectedMessage = "Error test \(1) \(NSNumber(value: 12.3))"
            XCTAssertEqual(message, expectedMessage)
            expectation.fulfill()
        }

        Logger.diagnostic(String(describing: TestLogger.self), message: String(format: "Error test %d %@", 1, NSNumber(value: 12.3)), errorOrException: nil)
        wait(for: [expectation], timeout: 10)
    }

    func testDiagnosticTrackingWithError() {
        let expectation = XCTestExpectation()
        let raisedError = NSError(domain: NSURLErrorDomain, code: 400, userInfo: nil)

        let diagnostic = MockDiagnosticLogger()
        diagnostic.callback = { tag, message, error, exception in
            XCTAssertEqual(tag, String(describing: TestLogger.self))
            XCTAssertEqual(message, "Error test")
            XCTAssert(error as? NSError == raisedError)
            expectation.fulfill()
        }

        Logger.diagnostic(String(describing: TestLogger.self), message: "Error test", errorOrException: raisedError)
        wait(for: [expectation], timeout: 10)
    }

    func testDiagnosticTrackingWithException() {
        let expectation = XCTestExpectation()
        let raisedException = NSException(name: .invalidArgumentException, reason: nil, userInfo: nil)

        let diagnostic = MockDiagnosticLogger()
        diagnostic.callback = { tag, message, error, exception in
            XCTAssertEqual(tag, String(describing: TestLogger.self))
            XCTAssertEqual(message, "Exception test")
            XCTAssertEqual(exception, raisedException)
            expectation.fulfill()
        }

        Logger.diagnostic(String(describing: TestLogger.self), message: "Exception test", errorOrException: raisedException)
        wait(for: [expectation], timeout: 10)
    }
}
